processor/src/main/java/org/derive4j/processor/derivator/LazyConstructorDerivator.java

```diff
@@ -73,20 +73,20 @@
                 .addModifiers(Modifier.PRIVATE)
                 .returns(typeName)
                 .addCode(CodeBlock.builder()
-                    .addStatement("$T _evaluation = this.evaluation", typeName)
-                    .beginControlFlow("if (_evaluation == null)")
-                    .beginControlFlow("synchronized (this.lock)")
```

Comment:
Good point, I was basing this on Suppliers.memoize, but I did not run any benchmarks — they seem to be trying to avoid making evaluation a volatile but I have no idea if that makes a difference in practice. I will see if I can come up with a benchmark to compare the different implementations.

> On Feb 13, 2016, at 10:51 AM, JB Giraudeau <notifications@github.com> wrote:
> 
> In processor/src/main/java/org/derive4j/processor/derivator/LazyConstructorDerivator.java <https://github.com/derive4j/derive4j/pull/29#discussion_r52831894>:
> 
> > @@ -73,20 +73,20 @@
> >                  .addModifiers(Modifier.PRIVATE)
> >                  .returns(typeName)
> >                  .addCode(CodeBlock.builder()
> > -                    .addStatement("$T _evaluation = this.evaluation", typeName)
> > -                    .beginControlFlow("if (_evaluation == null)")
> > -                    .beginControlFlow("synchronized (this.lock)")
> I did not synchronizing on this to avoid potential synchronization issues as the lazy value could potentially be used as a monitor. But then this would probably be bad practice. As a reference, I think Guava Suppliers.memoize does synchronize on this. I need to benchmark if avoiding the creation of the lock Object result in significative performance improvement.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/derive4j/derive4j/pull/29/files#r52831894>.
>